// @parcel/cache/lib/Cache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createCacheDir = createCacheDir;
exports.default = void 0;

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _logger() {
  const data = _interopRequireDefault(require("@parcel/logger"));

  _logger = function () {
    return data;
  };

  return data;
}

function _core() {
  const data = require("@parcel/core");

  _core = function () {
    return data;
  };

  return data;
}

function _package() {
  const data = _interopRequireDefault(require("../package.json"));

  _package = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Cache {
  constructor(fs, cacheDir) {
    _defineProperty(this, "fs", void 0);

    _defineProperty(this, "dir", void 0);

    this.fs = fs;
    this.dir = cacheDir;
  }

  _getCachePath(cacheId, extension = '.v8') {
    return _path().default.join(this.dir, cacheId.slice(0, 2), cacheId.slice(2) + extension);
  }

  getStream(key) {
    return this.fs.createReadStream(this._getCachePath(key, '.blob'));
  }

  setStream(key, stream) {
    return new Promise((resolve, reject) => {
      stream.pipe(this.fs.createWriteStream(this._getCachePath(key, '.blob'))).on('error', reject).on('finish', () => resolve(key));
    });
  }

  blobExists(key) {
    return this.fs.exists(this._getCachePath(key, '.blob'));
  }

  getBlob(key) {
    return this.fs.readFile(this._getCachePath(key, '.blob'));
  }

  async setBlob(key, contents) {
    await this.fs.writeFile(this._getCachePath(key, '.blob'), contents);
    return key;
  }

  async get(key) {
    try {
      let data = await this.fs.readFile(this._getCachePath(key));
      return (0, _core().deserialize)(data);
    } catch (err) {
      if (err.code === 'ENOENT') {
        return null;
      } else {
        throw err;
      }
    }
  }

  async set(key, value) {
    try {
      let blobPath = this._getCachePath(key);

      let data = (0, _core().serialize)(value);
      await this.fs.writeFile(blobPath, data);
      return key;
    } catch (err) {
      _logger().default.error(err, '@parcel/cache');
    }
  }

}

exports.default = Cache;

async function createCacheDir(fs, dir) {
  // First, create the main cache directory if necessary.
  await fs.mkdirp(dir); // In parallel, create sub-directories for every possible hex value
  // This speeds up large caches on many file systems since there are fewer files in a single directory.

  let dirPromises = [];

  for (let i = 0; i < 256; i++) {
    dirPromises.push(fs.mkdirp(_path().default.join(dir, ('00' + i.toString(16)).slice(-2))));
  }

  await Promise.all(dirPromises);
}

(0, _core().registerSerializableClass)(`${_package().default.version}:Cache`, Cache);
