// @parcel/core/lib/requests/AssetRequest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createAssetRequest;

function _utils() {
  const data = require("@parcel/utils");

  _utils = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _ParcelConfigRequest() {
  const data = _interopRequireDefault(require("./ParcelConfigRequest"));

  _ParcelConfigRequest = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createAssetRequest(input) {
  return {
    type: 'asset_request',
    id: getId(input),
    run,
    input
  };
}

const type = 'asset_request';

function getId(input) {
  // eslint-disable-next-line no-unused-vars
  let {
    optionsRef,
    ...hashInput
  } = input;
  return (0, _utils().md5FromOrderedObject)({
    type,
    filePath: input.filePath,
    env: input.env.id,
    isSource: input.isSource,
    sideEffects: input.sideEffects,
    code: input.code,
    pipeline: input.pipeline,
    query: input.query ? (0, _utils().objectSortedEntries)(input.query) : null
  });
}

async function run({
  input,
  api,
  farm,
  invalidateReason
}) {
  api.invalidateOnFileUpdate(input.filePath);
  let start = Date.now();
  let {
    optionsRef,
    ...rest
  } = input;
  let {
    cachePath
  } = (0, _nullthrows().default)(await api.runRequest((0, _ParcelConfigRequest().default)()));
  let previousDevDepRequests = new Map(await Promise.all(api.getSubRequests().filter(req => req.type === 'dev_dep_request').map(async req => [req.id, (0, _nullthrows().default)(await api.getRequestResult(req.id))])));
  let request = { ...rest,
    invalidateReason,
    // Add invalidations to the request if a node already exists in the graph.
    // These are used to compute the cache key for assets during transformation.
    invalidations: api.getInvalidations().filter(invalidation => {
      // Filter out invalidation node for the input file itself.
      return invalidation.type !== 'file' || invalidation.filePath !== input.filePath;
    }),
    devDeps: new Map([...previousDevDepRequests.entries()].filter(([id]) => api.canSkipSubrequest(id)).map(([, req]) => [`${req.moduleSpecifier}:${req.resolveFrom}`, req.hash])),
    invalidDevDeps: await Promise.all([...previousDevDepRequests.entries()].filter(([id]) => !api.canSkipSubrequest(id)).flatMap(([, req]) => {
      var _req$additionalInvali;

      return [{
        moduleSpecifier: req.moduleSpecifier,
        resolveFrom: req.resolveFrom
      }, ...((_req$additionalInvali = req.additionalInvalidations) !== null && _req$additionalInvali !== void 0 ? _req$additionalInvali : [])];
    }))
  };
  let {
    assets,
    configRequests,
    invalidations,
    invalidateOnFileCreate,
    devDepRequests
  } = await farm.createHandle('runTransform')({
    configCachePath: cachePath,
    optionsRef,
    request
  });
  let time = Date.now() - start;

  for (let asset of assets) {
    asset.stats.time = time;
  }

  for (let invalidation of invalidateOnFileCreate) {
    api.invalidateOnFileCreate(invalidation);
  }

  for (let invalidation of invalidations) {
    switch (invalidation.type) {
      case 'file':
        api.invalidateOnFileUpdate(invalidation.filePath);
        api.invalidateOnFileDelete(invalidation.filePath);
        break;

      case 'env':
        api.invalidateOnEnvChange(invalidation.key);
        break;

      case 'option':
        api.invalidateOnOptionChange(invalidation.key);
        break;

      default:
        throw new Error(`Unknown invalidation type: ${invalidation.type}`);
    }
  }

  for (let devDepRequest of devDepRequests) {
    await api.runRequest({
      id: 'dev_dep_request:' + devDepRequest.moduleSpecifier + ':' + devDepRequest.hash,
      type: 'dev_dep_request',
      run: ({
        api
      }) => {
        for (let filePath of (0, _nullthrows().default)(devDepRequest.invalidateOnFileChange)) {
          api.invalidateOnFileUpdate(filePath);
          api.invalidateOnFileDelete(filePath);
        }

        for (let invalidation of (0, _nullthrows().default)(devDepRequest.invalidateOnFileCreate)) {
          api.invalidateOnFileCreate(invalidation);
        }

        api.storeResult({
          moduleSpecifier: devDepRequest.moduleSpecifier,
          resolveFrom: devDepRequest.resolveFrom,
          hash: devDepRequest.hash,
          additionalInvalidations: devDepRequest.additionalInvalidations
        });
      },
      input: null
    });
  } // Add config requests


  for (let configRequest of configRequests) {
    await api.runRequest({
      id: 'config_request:' + configRequest.id,
      type: 'config_request',
      run: ({
        api
      }) => {
        let {
          includedFiles,
          invalidateOnFileCreate,
          shouldInvalidateOnStartup
        } = configRequest;

        for (let filePath of includedFiles) {
          api.invalidateOnFileUpdate(filePath);
          api.invalidateOnFileDelete(filePath);
        }

        for (let invalidation of invalidateOnFileCreate) {
          api.invalidateOnFileCreate(invalidation);
        }

        if (shouldInvalidateOnStartup) {
          api.invalidateOnStartup();
        }
      },
      input: null
    });
  }

  return assets;
}
