// @parcel/core/lib/Graph.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapVisitor = mapVisitor;
exports.default = exports.ALL_EDGE_TYPES = void 0;

function _assert() {
  const data = _interopRequireDefault(require("assert"));

  _assert = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const ALL_EDGE_TYPES = '@@all_edge_types';
exports.ALL_EDGE_TYPES = ALL_EDGE_TYPES;

class Graph {
  constructor(opts = {}) {
    _defineProperty(this, "nodes", void 0);

    _defineProperty(this, "inboundEdges", void 0);

    _defineProperty(this, "outboundEdges", void 0);

    _defineProperty(this, "rootNodeId", void 0);

    this.nodes = opts.nodes || new Map();
    this.rootNodeId = opts.rootNodeId;
    let edges = opts.edges;

    if (edges != null) {
      this.inboundEdges = new AdjacencyList(edges.inboundEdges);
      this.outboundEdges = new AdjacencyList(edges.outboundEdges);
    } else {
      this.inboundEdges = new AdjacencyList();
      this.outboundEdges = new AdjacencyList();
    }
  }

  static deserialize(opts) {
    return new this({
      nodes: opts.nodes,
      edges: opts.edges,
      rootNodeId: opts.rootNodeId
    });
  }

  serialize() {
    return {
      nodes: this.nodes,
      edges: {
        inboundEdges: this.inboundEdges.getListMap(),
        outboundEdges: this.outboundEdges.getListMap()
      },
      rootNodeId: this.rootNodeId
    };
  } // Returns a list of all edges in the graph. This can be large, so iterating
  // the complete list can be costly in large graphs. Used when merging graphs.


  getAllEdges() {
    let edges = [];

    for (let [from, edgeList] of this.outboundEdges.getListMap()) {
      for (let [type, toNodes] of edgeList) {
        for (let to of toNodes) {
          edges.push({
            from,
            to,
            type
          });
        }
      }
    }

    return edges;
  }

  addNode(node) {
    let existingNode = this.nodes.get(node.id);

    if (existingNode) {
      existingNode.value = node.value;
    }

    this.nodes.set(node.id, existingNode !== null && existingNode !== void 0 ? existingNode : node);
    return node;
  }

  hasNode(id) {
    return this.nodes.has(id);
  }

  getNode(id) {
    return this.nodes.get(id);
  }

  setRootNode(node) {
    this.addNode(node);
    this.rootNodeId = node.id;
  }

  getRootNode() {
    return this.rootNodeId ? this.getNode(this.rootNodeId) : null;
  }

  addEdge(from, to, type = null) {
    if (!this.getNode(from)) {
      throw new Error(`"from" node '${from}' not found`);
    }

    if (!this.getNode(to)) {
      throw new Error(`"to" node '${to}' not found`);
    }

    this.outboundEdges.addEdge(from, to, type);
    this.inboundEdges.addEdge(to, from, type);
  }

  hasEdge(from, to, type = null) {
    return this.outboundEdges.hasEdge(from, to, type);
  }

  getNodesConnectedTo(node, type = null) {
    assertHasNode(this, node);
    let inboundByType = this.inboundEdges.getEdgesByType(node.id);

    if (inboundByType == null) {
      return [];
    }

    let nodes;

    if (type === ALL_EDGE_TYPES) {
      nodes = new Set();

      for (let [, typeNodes] of inboundByType) {
        for (let node of typeNodes) {
          nodes.add(node);
        }
      }
    } else if (Array.isArray(type)) {
      nodes = new Set();

      for (let typeName of type) {
        for (let node of (_inboundByType$get$va = (_inboundByType$get = inboundByType.get(typeName)) === null || _inboundByType$get === void 0 ? void 0 : _inboundByType$get.values()) !== null && _inboundByType$get$va !== void 0 ? _inboundByType$get$va : []) {
          var _inboundByType$get$va, _inboundByType$get;

          nodes.add(node);
        }
      }
    } else {
      var _inboundByType$get$va2, _inboundByType$get2;

      nodes = new Set((_inboundByType$get$va2 = (_inboundByType$get2 = inboundByType.get(type)) === null || _inboundByType$get2 === void 0 ? void 0 : _inboundByType$get2.values()) !== null && _inboundByType$get$va2 !== void 0 ? _inboundByType$get$va2 : []);
    }

    return [...nodes].map(to => (0, _nullthrows().default)(this.nodes.get(to)));
  }

  getNodesConnectedFrom(node, type = null) {
    assertHasNode(this, node);
    let outboundByType = this.outboundEdges.getEdgesByType(node.id);

    if (outboundByType == null) {
      return [];
    }

    let nodes;

    if (type === ALL_EDGE_TYPES) {
      nodes = new Set();

      for (let [, typeNodes] of outboundByType) {
        for (let node of typeNodes) {
          nodes.add(node);
        }
      }
    } else if (Array.isArray(type)) {
      nodes = new Set();

      for (let typeName of type) {
        for (let node of (_outboundByType$get$v = (_outboundByType$get = outboundByType.get(typeName)) === null || _outboundByType$get === void 0 ? void 0 : _outboundByType$get.values()) !== null && _outboundByType$get$v !== void 0 ? _outboundByType$get$v : []) {
          var _outboundByType$get$v, _outboundByType$get;

          nodes.add(node);
        }
      }
    } else {
      var _outboundByType$get$v2, _outboundByType$get2;

      nodes = new Set((_outboundByType$get$v2 = (_outboundByType$get2 = outboundByType.get(type)) === null || _outboundByType$get2 === void 0 ? void 0 : _outboundByType$get2.values()) !== null && _outboundByType$get$v2 !== void 0 ? _outboundByType$get$v2 : []);
    }

    return [...nodes].map(to => (0, _nullthrows().default)(this.nodes.get(to)));
  } // Removes node and any edges coming from or to that node


  removeNode(node) {
    assertHasNode(this, node);

    for (let [type, nodesForType] of this.inboundEdges.getEdgesByType(node.id)) {
      for (let from of nodesForType) {
        this.removeEdge(from, node.id, type, // Do not allow orphans to be removed as this node could be one
        // and is already being removed.
        false
        /* removeOrphans */
        );
      }
    }

    for (let [type, toNodes] of this.outboundEdges.getEdgesByType(node.id)) {
      for (let to of toNodes) {
        this.removeEdge(node.id, to, type);
      }
    }

    let wasRemoved = this.nodes.delete(node.id);
    (0, _assert().default)(wasRemoved);
  }

  removeById(id) {
    let node = (0, _nullthrows().default)(this.getNode(id));
    this.removeNode(node);
  }

  removeEdges(node, type = null) {
    assertHasNode(this, node);

    for (let to of this.outboundEdges.getEdges(node.id, type)) {
      this.removeEdge(node.id, to, type);
    }
  } // Removes edge and node the edge is to if the node is orphaned


  removeEdge(from, to, type = null, removeOrphans = true) {
    if (!this.outboundEdges.hasEdge(from, to, type)) {
      throw new Error(`Outbound edge from ${from} to ${to} not found!`);
    }

    if (!this.inboundEdges.hasEdge(to, from, type)) {
      throw new Error(`Inbound edge from ${to} to ${from} not found!`);
    }

    this.outboundEdges.removeEdge(from, to, type);
    this.inboundEdges.removeEdge(to, from, type);
    let connectedNode = (0, _nullthrows().default)(this.nodes.get(to));

    if (removeOrphans && this.isOrphanedNode(connectedNode)) {
      this.removeNode(connectedNode);
    }
  }

  isOrphanedNode(node) {
    assertHasNode(this, node);
    let rootNode = this.getRootNode();

    if (rootNode == null) {
      // If the graph does not have a root, and there are inbound edges,
      // this node should not be considered orphaned.
      // return false;
      for (let [, inboundNodeIds] of this.inboundEdges.getEdgesByType(node.id)) {
        if (inboundNodeIds.size > 0) {
          return false;
        }
      }

      return true;
    } // Otherwise, attempt to traverse backwards to the root. If there is a path,
    // then this is not an orphaned node.


    let hasPathToRoot = false;
    this.traverseAncestors(node, (ancestor, _, actions) => {
      if (ancestor.id === rootNode.id) {
        hasPathToRoot = true;
        actions.stop();
      }
    }, // $FlowFixMe
    ALL_EDGE_TYPES);

    if (hasPathToRoot) {
      return false;
    }

    return true;
  }

  replaceNode(fromNode, toNode, type = null) {
    assertHasNode(this, fromNode);
    this.addNode(toNode);

    for (let parent of this.inboundEdges.getEdges(fromNode.id, type)) {
      this.addEdge(parent, toNode.id, type);
      this.removeEdge(parent, fromNode.id, type);
    }

    this.removeNode(fromNode);
  } // Update a node's downstream nodes making sure to prune any orphaned branches


  replaceNodesConnectedTo(fromNode, toNodes, replaceFilter, type = null) {
    assertHasNode(this, fromNode);
    let outboundEdges = this.outboundEdges.getEdges(fromNode.id, type);
    let childrenToRemove = new Set(replaceFilter ? [...outboundEdges].filter(toNodeId => replaceFilter((0, _nullthrows().default)(this.nodes.get(toNodeId)))) : outboundEdges);

    for (let toNode of toNodes) {
      this.addNode(toNode);
      childrenToRemove.delete(toNode.id);

      if (!this.hasEdge(fromNode.id, toNode.id, type)) {
        this.addEdge(fromNode.id, toNode.id, type);
      }
    }

    for (let child of childrenToRemove) {
      this.removeEdge(fromNode.id, child, type);
    }
  }

  traverse(visit, startNode, type = null) {
    return this.dfs({
      visit,
      startNode,
      getChildren: node => this.getNodesConnectedFrom(node, type)
    });
  }

  filteredTraverse(filter, visit, startNode, type) {
    return this.traverse(mapVisitor(filter, visit), startNode, type);
  }

  traverseAncestors(startNode, visit, type = null) {
    return this.dfs({
      visit,
      startNode,
      getChildren: node => this.getNodesConnectedTo(node, type)
    });
  }

  dfs({
    visit,
    startNode,
    getChildren
  }) {
    let root = startNode !== null && startNode !== void 0 ? startNode : this.getRootNode();

    if (root == null) {
      throw new Error('A start node is required to traverse');
    }

    assertHasNode(this, root);
    let visited = new Set();
    let stopped = false;
    let skipped = false;
    let actions = {
      skipChildren() {
        skipped = true;
      },

      stop() {
        stopped = true;
      }

    };

    let walk = (node, context) => {
      if (!this.hasNode(node.id)) return;
      visited.add(node);
      skipped = false;
      let enter = typeof visit === 'function' ? visit : visit.enter;

      if (enter) {
        let newContext = enter(node, context, actions);

        if (typeof newContext !== 'undefined') {
          context = newContext;
        }
      }

      if (skipped) {
        return;
      }

      if (stopped) {
        return context;
      }

      for (let child of getChildren(node)) {
        if (visited.has(child)) {
          continue;
        }

        visited.add(child);
        let result = walk(child, context);

        if (stopped) {
          return result;
        }
      }

      if (typeof visit !== 'function' && visit.exit) {
        let newContext = visit.exit(node, context, actions);

        if (typeof newContext !== 'undefined') {
          context = newContext;
        }
      }

      if (skipped) {
        return;
      }

      if (stopped) {
        return context;
      }
    };

    return walk(root);
  }

  bfs(visit) {
    let root = this.getRootNode();

    if (!root) {
      throw new Error('A root node is required to traverse');
    }

    let queue = [root];
    let visited = new Set([root]);

    while (queue.length > 0) {
      let node = queue.shift();
      let stop = visit(node);

      if (stop === true) {
        return node;
      }

      for (let child of this.getNodesConnectedFrom(node)) {
        if (!visited.has(child)) {
          visited.add(child);
          queue.push(child);
        }
      }
    }

    return null;
  }

  getSubGraph(node) {
    let graph = new this.constructor();
    graph.setRootNode(node);
    let nodes = [];
    this.traverse(node => {
      nodes.push(node);
      graph.addNode(node);
    }, node);

    for (let node of nodes) {
      for (let [type, toNodes] of this.outboundEdges.getEdgesByType(node.id)) {
        for (let to of toNodes) {
          graph.addEdge(node.id, to, type);
        }
      }
    }

    return graph;
  }

  findAncestor(node, fn) {
    let res = null;
    this.traverseAncestors(node, (node, ctx, traversal) => {
      if (fn(node)) {
        res = node;
        traversal.stop();
      }
    });
    return res;
  }

  findAncestors(node, fn) {
    let res = [];
    this.traverseAncestors(node, (node, ctx, traversal) => {
      if (fn(node)) {
        res.push(node);
        traversal.skipChildren();
      }
    });
    return res;
  }

  findDescendant(node, fn) {
    let res = null;
    this.traverse((node, ctx, traversal) => {
      if (fn(node)) {
        res = node;
        traversal.stop();
      }
    }, node);
    return res;
  }

  findDescendants(node, fn) {
    let res = [];
    this.traverse((node, ctx, traversal) => {
      if (fn(node)) {
        res.push(node);
        traversal.skipChildren();
      }
    }, node);
    return res;
  }

  findNode(predicate) {
    return Array.from(this.nodes.values()).find(predicate);
  }

  findNodes(predicate) {
    return Array.from(this.nodes.values()).filter(predicate);
  }

}

exports.default = Graph;

function mapVisitor(filter, visit) {
  return {
    enter: (node, context, actions) => {
      let enter = typeof visit === 'function' ? visit : visit.enter;

      if (!enter) {
        return;
      }

      let value = filter(node, actions);

      if (value != null) {
        return enter(value, context, actions);
      }
    },
    exit: (node, context, actions) => {
      if (typeof visit === 'function') {
        return;
      }

      let exit = visit.exit;

      if (!exit) {
        return;
      }

      let value = filter(node, actions);

      if (value != null) {
        return exit(value, context, actions);
      }
    }
  };
}

function assertHasNode(graph, node) {
  if (!graph.hasNode(node.id)) {
    throw new Error('Does not have node ' + node.id);
  }
}

class AdjacencyList {
  constructor(listMap) {
    _defineProperty(this, "_listMap", void 0);

    this._listMap = listMap !== null && listMap !== void 0 ? listMap : new Map();
  }

  getListMap() {
    return this._listMap;
  }

  getEdges(from, type) {
    var _this$_listMap$get$ge, _this$_listMap$get;

    return (_this$_listMap$get$ge = (_this$_listMap$get = this._listMap.get(from)) === null || _this$_listMap$get === void 0 ? void 0 : _this$_listMap$get.get(type)) !== null && _this$_listMap$get$ge !== void 0 ? _this$_listMap$get$ge : new Set();
  }

  getEdgesByType(from) {
    var _this$_listMap$get2;

    return (_this$_listMap$get2 = this._listMap.get(from)) !== null && _this$_listMap$get2 !== void 0 ? _this$_listMap$get2 : new Map();
  }

  hasEdge(from, to, type) {
    var _this$_listMap$get3, _this$_listMap$get3$g;

    return Boolean((_this$_listMap$get3 = this._listMap.get(from)) === null || _this$_listMap$get3 === void 0 ? void 0 : (_this$_listMap$get3$g = _this$_listMap$get3.get(type)) === null || _this$_listMap$get3$g === void 0 ? void 0 : _this$_listMap$get3$g.has(to));
  }

  addEdge(from, to, type) {
    let types = this._listMap.get(from);

    if (types == null) {
      types = new Map();

      this._listMap.set(from, types);
    }

    let adjacent = types.get(type);

    if (adjacent == null) {
      adjacent = new Set();
      types.set(type, adjacent);
    }

    adjacent.add(to);
  }

  removeEdge(from, to, type) {
    var _this$_listMap$get4, _this$_listMap$get4$g;

    (_this$_listMap$get4 = this._listMap.get(from)) === null || _this$_listMap$get4 === void 0 ? void 0 : (_this$_listMap$get4$g = _this$_listMap$get4.get(type)) === null || _this$_listMap$get4$g === void 0 ? void 0 : _this$_listMap$get4$g.delete(to);
  }

}
